### PR TITLE
[SPARK-11269][SQL] Java API support & test cases

### PR DESCRIPTION
**Pull request by cloud-fan**
I followed `DataFrame` to add java friendly methods in `Dataset`.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala [hunk 1]
```diff
@@ -37,3 +35,39 @@ trait Encoder[T] extends Serializable {
...
+  def forShort: Encoder[java.lang.Short] = ExpressionEncoder(flat = true)
+  def forInt: Encoder[java.lang.Integer] = ExpressionEncoder(flat = true)
+  def forLong: Encoder[java.lang.Long] = ExpressionEncoder(flat = true)
+  def forFloat: Encoder[java.lang.Float] = ExpressionEncoder(flat = true)
+  def forDouble: Encoder[java.lang.Double] = ExpressionEncoder(flat = true)
+  def forString: Encoder[java.lang.String] = ExpressionEncoder(flat = true)
+
+  def typeTagOfTuple2[T1 : TypeTag, T2 : TypeTag]: TypeTag[(T1, T2)] = typeTag[(T1, T2)]
+
+  private def getTypeTag[T](c: Class[T]): TypeTag[T] = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala]: This is really really hacky, I'm not sure if I did it right...

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+  public void tearDown() {
+    context.sparkContext().stop();
+    context = null;
+    jsc = null;
+  }
+
+  @Test
+  public void testCommonOperation() {
+    List<String> data = Arrays.asList("hello", "world");
+    Dataset<String> ds = context.createDataset(data, Encoder$.MODULE$.forString());
```
Comment (cloud-fan) [inline sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java]: `Encoder$.MODULE$` looks very weird to me, is there a better way to use scala object in java?

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 3]
```diff
@@ -409,6 +443,9 @@ class Dataset[T] private(
   /** Collects the elements to an Array. */
   def collect(): Array[T] = rdd.collect()
 
+  def jcollect(): java.util.List[T] =
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: I do want to avoid this `jcollect`, but our `collect` with return type `Array[T]` seems not working at java side, any ideas?

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala [hunk 4]
```diff
@@ -499,6 +499,10 @@ class SQLContext private[sql](
     new Dataset[T](this, plan)
   }
 
+  def createDataset[T : Encoder](data: java.util.List[T]): Dataset[T] = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: I found that we can't create `Dataset` from `RDD`, should we add it?

**Comment by cloud-fan** [comment]
Now most of our `Dataset` APIs can work at java side, I'm waiting for some feedback on my approach so that I can improve it(e.g. how to create encoder for tuples).

**Comment by cloud-fan** [comment]
ping @marmbrus

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 3 above)
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: That `Array` should be a java array? Why it does not work?

**Comment by cloud-fan** [comment]
cc @yhuai @marmbrus  , according to the jenkins result, scala method which returns generic array will return `Object` at java side, `DataFrame.collect` returns `Array[Row]` and has no problem.

The reason is that: in scala `Array[Int]`, `Array[String]` are just `Array` with type parameter, however, at java side, `int[]` and `String[]` is different type, and there is exactly one supertype of them: `java.lang.Object`.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 5]
```diff
@@ -441,6 +537,17 @@ class Dataset[T] private(
   /** Collects the elements to an Array. */
   def collect(): Array[T] = rdd.collect()
 
+  /**
+   * (Java-specific)
+   * Collects the elements to a Java list.
+   *
+   * Due to the incompatibility problem between Scala and Java, the return type of [[collect()]] at
+   * Java side is `java.lang.Object`, which is not easy to use.  Java user can use this method
+   * instead and keep the generic type for result.
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: @since

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala (hunk 4 above)
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: Yeah, lets do that in another PR please.

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: Oh, you already did it :)

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 6]
```diff
@@ -441,6 +537,17 @@ class Dataset[T] private(
   /** Collects the elements to an Array. */
   def collect(): Array[T] = rdd.collect()
 
+  /**
+   * (Java-specific)
+   * Collects the elements to a Java list.
+   *
+   * Due to the incompatibility problem between Scala and Java, the return type of [[collect()]] at
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: This just means that the RDD has the wrong classtag.  We need to find a way to pass the classtag from the encoder before calling collect.

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java (hunk 2 above)
Comment (marmbrus) [inline sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java]: Yeah, scala should create static methods so that you can just call `Encoder.forString()`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala [hunk 7]
```diff
@@ -37,3 +35,39 @@ trait Encoder[T] extends Serializable {
...
+    val tpe = sym.selfType
+    TypeTag(mirror, new api.TypeCreator {
+      def apply[U <: api.Universe with Singleton](m: api.Mirror[U]) =
+        if (m eq mirror) tpe.asInstanceOf[U # Type]
+        else throw new IllegalArgumentException(
+          s"Type tag defined in $mirror cannot be migrated to other mirrors.")
+    })
+  }
+
+  def forTuple2[T1, T2](c1: Class[T1], c2: Class[T2]): Encoder[(T1, T2)] = {
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala]: How about just `forTuple`, the type of tuple returned is obvious from the number of arguments.  We should also add at least up to tuple 5.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala [hunk 8]
```diff
@@ -37,3 +35,39 @@ trait Encoder[T] extends Serializable {
...
+  def forBoolean: Encoder[java.lang.Boolean] = ExpressionEncoder(flat = true)
+  def forByte: Encoder[java.lang.Byte] = ExpressionEncoder(flat = true)
+  def forShort: Encoder[java.lang.Short] = ExpressionEncoder(flat = true)
+  def forInt: Encoder[java.lang.Integer] = ExpressionEncoder(flat = true)
+  def forLong: Encoder[java.lang.Long] = ExpressionEncoder(flat = true)
+  def forFloat: Encoder[java.lang.Float] = ExpressionEncoder(flat = true)
+  def forDouble: Encoder[java.lang.Double] = ExpressionEncoder(flat = true)
+  def forString: Encoder[java.lang.String] = ExpressionEncoder(flat = true)
+
+  def typeTagOfTuple2[T1 : TypeTag, T2 : TypeTag]: TypeTag[(T1, T2)] = typeTag[(T1, T2)]
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala]: `private`

**Comment by marmbrus** [comment]
It would be really great to also try and create a test suite that uses java 8 lambdas (though we may need to pull this into a separate PR as I'm not sure how many build changes we will need)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala [hunk 9]
```diff
@@ -37,3 +35,39 @@ trait Encoder[T] extends Serializable {
   /** A ClassTag that can be used to construct and Array to contain a collection of `T`. */
   def clsTag: ClassTag[T]
 }
+
+object Encoder {
+  import scala.reflect.runtime.universe._
+
+  def forBoolean: Encoder[java.lang.Boolean] = ExpressionEncoder(flat = true)
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala]: @rxin @mateiz thoughts on naming here?

`forInt` `int` `INT`?

`forTuple` `tuple`, `tuple2`?

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala]: Encoder.int seems simpler.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 6 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Will the class tag do the trick? I tried to define a generic class with ClassTag:
```
class MyTest[T : ClassTag] {
  def t(): Array[T] = null
}

object MyTest {
  def apply[T](cls: Class[T]): MyTest[T] = {
    new MyTest[T]()(ClassTag(cls))
  }
}
```

The return type of `MyClass.t()` is still `Object` at java side.
I also tried to use scala RDD at java side, the return type of `RDD.collect()` is also `Object`.

One possible solution is to define `T <: AnyRef`, but I think it's hard to make it for `Dataset` or `RDD`.

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: RDD holds a class tag of the element type that it uses to construct the
correct type of array when you do a collect.

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: cc @rxin

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: We can construct right type of array while calling `RDD.collect`, the problem is the interface. At java side the return type of `RDD.collect()` is `java.lang.Object` and we need to do a type cast, which is not friendly to users.

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: I see, then we should have collect as list too.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala (hunk 9 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala]: Actually we cannot, as `int` is a keyword in java. Maybe `INT`?

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala]: INT sounds good...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala [hunk 10]
```diff
@@ -37,3 +37,120 @@ trait Encoder[T] extends Serializable {
...
+      enc1: Encoder[T1],
+      enc2: Encoder[T2],
+      enc3: Encoder[T3],
+      enc4: Encoder[T4],
+      enc5: Encoder[T5]): Encoder[(T1, T2, T3, T4, T5)] = {
+    tuple(Seq(enc1, enc2, enc3, enc4, enc5).map(_.asInstanceOf[ExpressionEncoder[_]]))
+      .asInstanceOf[ExpressionEncoder[(T1, T2, T3, T4, T5)]]
+  }
+
+  private def tuple(encoders: Seq[ExpressionEncoder[_]]): ExpressionEncoder[_] = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala]: I came out with a new approach to create encoder at java side, which can also support nested tuples as well. I also keep the old code below and you can judge which way is better.

cc @marmbrus  @rxin

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 11]
```diff
@@ -17,9 +17,13 @@
 
 package org.apache.spark.sql
 
+import scala.collection.JavaConverters._
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: let's use explicit conversions instead of implicit ones.

Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: JavaConverters _is_ the explicit one (`.asScala` / `.asJava`); the more implicit one was banned by me in a Scalastyle updtae.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 12]
```diff
@@ -148,18 +152,37 @@ class Dataset[T] private(
...
    * @since 1.6.0
    */
   def filter(func: T => Boolean): Dataset[T] = mapPartitions(_.filter(func))
 
   /**
+   * (Java-specific)
+   * Returns a new [[Dataset]] that only contains elements where `func` returns `true`.
+   * @since 1.6.0
+   */
+  def filter(func: JFunction[T, java.lang.Boolean]): Dataset[T] =
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: After talking with @rxin we should probably create `FilterFunction` both because here we can avoid boxing and also because this might be less confusing to java users.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 13]
```diff
@@ -148,18 +152,37 @@ class Dataset[T] private(
...
    * @since 1.6.0
    */
   def map[U : Encoder](func: T => U): Dataset[U] = mapPartitions(_.map(func))
 
   /**
+   * (Java-specific)
+   * Returns a new [[Dataset]] that contains the result of applying `func` to each element.
+   * @since 1.6.0
+   */
+  def map[U](func: JFunction[T, U], encoder: Encoder[U]): Dataset[U] =
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Same here, `MapFunction`

**Comment by rxin** [comment]
@cloud-fan can you close this one? I'm going to merge https://github.com/apache/spark/pull/9528 for you

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala (hunk 10 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala]: @cloud-fan , does that mean the limit will be 22? Do you think we should at least add it up to Tuple22, which is the limit of Scala?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala (hunk 7 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala]: @marmbrus Any reason why it is tuple 5, instead of tuple 22 which is the current limit of Scala?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala (hunk 10 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala]: We can hold it off until some use cases come out that need more than Tuple5.

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala]: Thank you!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala [hunk 14]
```diff
@@ -37,3 +37,120 @@ trait Encoder[T] extends Serializable {
...
+  import scala.reflect.runtime.universe._
+
+  def BOOLEAN: Encoder[java.lang.Boolean] = ExpressionEncoder(flat = true)
+  def BYTE: Encoder[java.lang.Byte] = ExpressionEncoder(flat = true)
+  def SHORT: Encoder[java.lang.Short] = ExpressionEncoder(flat = true)
+  def INT: Encoder[java.lang.Integer] = ExpressionEncoder(flat = true)
+  def LONG: Encoder[java.lang.Long] = ExpressionEncoder(flat = true)
+  def FLOAT: Encoder[java.lang.Float] = ExpressionEncoder(flat = true)
+  def DOUBLE: Encoder[java.lang.Double] = ExpressionEncoder(flat = true)
+  def STRING: Encoder[java.lang.String] = ExpressionEncoder(flat = true)
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala]: @cloud-fan Could you share me your idea why we do not add the other basic types like DecimalType, DateType and TimestampType? Thank you! 

DecimalType -> java.math.BigDecimal
DateType -> java.sql.Date
TimestampType -> java.sql.Timestamp

Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala]: We should add these.

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala]: Will do it soon. Thanks!
